src/firmware/main/app/control.c

```diff
@@ -0,0 +1,182 @@
+    float prev_linear_accel_x           = controller_state->last_applied_acceleration_x;
+    float prev_linear_accel_y           = controller_state->last_applied_acceleration_y;
+    float prev_angular_accel = controller_state->last_applied_acceleration_angular;
+
+    float linear_diff_x = linear_accel_x - prev_linear_accel_x;
+    float linear_diff_y = linear_accel_y - prev_linear_accel_y;
+    float angular_diff  = angular_accel - prev_angular_accel;
+
+    const float jerk_limit         = robot_constants.jerk_limit;
+    const float linear_jerk_limit  = robot_constants.jerk_limit * TICK_TIME;
+    const float angular_jerk_limit = jerk_limit / ROBOT_RADIUS * TICK_TIME * 5.0f;
+    limit(&linear_diff_x, linear_jerk_limit);
+    limit(&linear_diff_y, linear_jerk_limit);
+    limit(&angular_diff, angular_jerk_limit);
+
+    linear_accel_x = prev_linear_accel_x + linear_diff_x;
+    linear_accel_y = prev_linear_accel_y + linear_diff_y;
+    angular_accel  = prev_angular_accel + angular_diff;
+
+    controller_state->last_applied_acceleration_x       = linear_accel_x;
+    controller_state->last_applied_acceleration_y       = linear_accel_y;
+    controller_state->last_applied_acceleration_angular = angular_accel;
+
+    float robot_force[3];
+    robot_force[0] = linear_accel_x * robot_constants.mass;
+    robot_force[1] = linear_accel_y * robot_constants.mass;
+    // input is angular acceleration so mass * Radius * radians/second^2 gives newtons
+    robot_force[2] =
+        angular_accel * robot_constants.robot_radius * robot_constants.moment_of_inertia;
+    float wheel_force[4];
+    speed3_to_speed4(robot_force, wheel_force);  // Convert to wheel coordinate syste
+
+    app_wheel_applyForce(app_firmware_robot_getFrontLeftWheel(robot), wheel_force[0]);
+    app_wheel_applyForce(app_firmware_robot_getFrontRightWheel(robot), wheel_force[3]);
+    app_wheel_applyForce(app_firmware_robot_getBackLeftWheel(robot), wheel_force[1]);
+    app_wheel_applyForce(app_firmware_robot_getBackRightWheel(robot), wheel_force[2]);
+}
+
+void app_control_trackVelocity(FirmwareRobot_t* robot, float linear_velocity_x,
+                               float linear_velocity_y, float angular_velocity)
+{
+    float current_vx               = app_firmware_robot_getVelocityX(robot);
+    float current_vy               = app_firmware_robot_getVelocityY(robot);
+    float current_angular_velocity = app_firmware_robot_getVelocityAngular(robot);
+    float current_orientation      = app_firmware_robot_getOrientation(robot);
+
+
+    // This is the "P" term in a PID controller. We essentially do proportional
+    // control of our acceleration based on velocity error
+    static const float ACCELERATION_GAIN = 10.0f;
+
+    float current_acceleration[2];
+    current_acceleration[0] = (linear_velocity_x - current_vx) * ACCELERATION_GAIN;
+    current_acceleration[1] = (linear_velocity_y - current_vy) * ACCELERATION_GAIN;
```

Comment:
Not a big fan of the naming here.

current_acceleration is not 'current' per se, it's more desired.
ACCELERATION_GAIN is not scaling acceleration, it's more of a velocity error gain.

Reply:
you're totally right, renamed `current_acceleration` -> `desired_acceleration`, `ACCELERATION_GAIN`->`VELOCITY_ERROR_GAIN`